### ISSUE TITLE
Cygwin doesn't work with Git Bash but does with Windows Shell

### ISSUE THREAD

**Issue by embcla** [git-for-windows/git#3728]
- [x] I was not able to find an [open](https://github.com/git-for-windows/git/issues?q=is%3Aopen) or [closed](https://github.com/git-for-windows/git/issues?q=is%3Aclosed) issue matching what I'm seeing

I am installing a new Win11 laptop, and what I see is that I can't get Git Bash to cooperate with Cygwin.
I installed both as 64bit versions.

Cygwin commands (in this bug exemplified by Python3.9) run fine in windows shell, but cause a heap mismatch in Git Bash.
See attached screenshot.
<img width="1385" alt="image" src="https://user-images.githubusercontent.com/54371771/157647739-faa572af-7d90-4629-884b-319efac39bf6.png">

I really tried everything I could before opening this, but can't really figure this one out.
Previously I used Cygwin32, but it doesn't seem to be available anymore.
Disabled ASLR but made no difference.
Made sure there is only 1 cygwin1.dll on path.
Reinstalled multiple times but still have the same issue.
I think it's Git Bash because everything else works from Windows, so I believe that Git Bash is mangling some linker code and causing this issue.

### Setup

 - Which version of Git for Windows are you using? Is it 32-bit or 64-bit?

```
$ git --version --build-options

git version 2.35.1.windows.2
cpu: x86_64
built from commit: 5437f0fd368c7faf1a0b5e1fef048232c1f2a3e6
sizeof-long: 4
sizeof-size_t: 8
shell-path: /bin/sh
feature: fsmonitor--daemon
```

 - Which version of Windows are you running? Vista, 7, 8, 10? Is it 32-bit or 64-bit?

```
$ cmd.exe /c ver

Microsoft Windows [Version 10.0.22000.556]
```

 - What options did you set as part of the installation? Or did you choose the
   defaults?

```
> type "C:\Program Files\Git\etc\install-options.txt"

Editor Option: VisualStudioCode
Custom Editor Path:
Default Branch Option:
Path Option: Cmd
SSH Option: OpenSSH
Tortoise Option: false
CURL Option: OpenSSL
CRLF Option: CRLFAlways
Bash Terminal Option: MinTTY
Git Pull Behavior Option: Rebase
Use Credential Manager: Enabled
Performance Tweaks FSCache: Enabled
Enable Symlinks: Disabled
Enable Pseudo Console Support: Disabled
Enable FSMonitor: Disabled
```

 - Any other interesting things about your environment that might be related
   to the issue you're seeing?

Not sure

### Details

 - Which terminal/shell are you running Git from? e.g Bash/CMD/PowerShell/other

Git Bash

 - What commands did you run to trigger this issue? If you can provide a
   [Minimal, Complete, and Verifiable example](http://stackoverflow.com/help/mcve)
   this will help us understand the issue.

```
python3.9
```
 - What did you expect to occur after running these commands?

That I'd get into the python interpreter

 - What actually happened instead?

Segfault
```
$ python3.9
      0 [main] python3.9 (22568) C:\cygwin64\bin\python3.9.exe: *** fatal error - cygheap base mismatch detected - 0x180349408/0x180346408.
This problem is probably due to using incompatible versions of the cygwin DLL.
Search for cygwin1.dll using the Windows Start->Find/Search facility
and delete all but the most recent version.  The most recent version *should*
reside in x:\cygwin\bin, where 'x' is the drive on which you have
installed the cygwin distribution.  Rebooting is also suggested if you
are unable to find another cygwin DLL.
```

**Comment by dscho** [comment]
> Search for cygwin1.dll using the Windows Start->Find/Search facility

Sadly, the problem is not a duplicate `cygwin1.dll` but instead `msys-2.0.dll`, which is a friendly fork of Cygwin. Both DLLs implement a POSIX emulation layer (most notably to emulate `fork()` and pseudo terminals), and they assume that one is compatible with the other, which they aren't.

But even if you could fix those compatibility problems _somehow_, the problems would not end there. Internally, Cygwin and MSYS2 are _just_ different enough not to work well with one another. For example, Cygwin's `less.exe` is unlikely to work well with MSYS2's pseudo terminals, and vice versa.

My best suggestion is to decouple the two POSIX emulation layers by running the Cygwin Python via `winpty`:

```console
$ winpty python3.9
```

But then, I wonder why you want to call Cygwin from Git for Windows? We _kind of_ support the other way round, at least a _little bit_.

Having said that, I always strongly recommend to stay within your favorite ecosystem. For example, why not use Cygwin's Bash and Cygwin's Git when you also want to use Cygwin's Python?

**Comment by embcla** [comment]
So I was using this system because Cygwin allows me to choose the packages I use on windows (gcc, meson, make, etc).

And to be honest I was just firing Git Bash that got me into a terminal, I did not even know Cygwin came with its own terminal, now I checked and Indeed it's there.

Also on my previous laptop, this just worked: Git Bash provided my main bash environment, and then I'd use Cygwin to install and update packages.


So, for me to better understand: which package is loading and installing MSYS2, Git for Windows?
